common/tools/dev-tool/src/util/commandBuilder.ts

```diff
@@ -0,0 +1,220 @@
+ * fully dynamic map will be produced.
+ */
+export type ParsedOptions<Opts extends CommandOptions | undefined> = {
+  /**
+   * If the argument "--" was encountered when parsing, this property
+   * holds all arguments that occurred after the "--"
+   */
+  "--": string[] | undefined;
+  /**
+   * Array of arguments to the command
+   */
+  args: string[];
+} & (Opts extends CommandOptions
+  ? {
+      // This heinous type signature here just takes the options structure
+      // and folds it into a map from long-key to resulting type.
+      // If no default is specified, then the type will be implicitly or'd
+      // with undefined
+      [K in Exclude<keyof Opts, "--" | "args" | "help">]: undefined extends Opts[K]["default"]
+        ?
+            | (Opts[K]["kind"] extends "string"
+                ? string
+                : Opts[K]["kind"] extends "boolean"
+                ? boolean
+                : Opts[K]["kind"] extends "multistring"
+                ? string[]
+                : string[] | string | boolean)
+            | undefined
+        : Opts[K]["kind"] extends "string"
+        ? string
+        : Opts[K]["kind"] extends "boolean"
+        ? boolean
+        : Opts[K]["kind"] extends "multistring"
+        ? string[]
+        : string[] | string | boolean;
+    }
+  : { [k: string]: any });
+
+/**
+ * Helper function to convert an args array to a parsed argument map.
+ *
+ * This is a strongly-typed wrapper around `minimist`.
+ *
+ * You should probably not use this function directly. The
+ * options should be parsed and their types should be inferred
+ * from {@link leafCommand}.
+ *
+ * @param args input array of arg strings from argv
+ * @param options option map of a command
+ */
+export function parseOptions<Opts extends CommandOptions>(
+  args: string[],
+  opts?: Opts
+): ParsedOptions<Opts> {
+  // If options are not provided, use an empty set
+  const options: CommandOptions = opts ?? {};
+
+  const keys = Object.keys(options).filter((k) => hasKey(options, k));
+  const argMap = getArgs(args, {
+    // Once an unidentified argument is encountered, stop parsing
+    stopEarly: true,
+    // Use type information for hinting to minimist about how arguments should
+    // be handled
+    boolean: ["help", ...keys.filter((k) => options[k].kind === "boolean")],
+    string: keys.filter((k) => options[k].kind === "string"),
+    // Roll up the optional short-names into aliases
+    alias: keys.reduce(
+      (o, key) =>
+        options[key].shortName !== undefined ? { ...o, [options[key].shortName!]: key } : o,
+      {}
+    ),
+    // Roll up the default values into the arg parser
+    default: {
+      ...keys.reduce(
+        (o, key) =>
+          options[key].default !== undefined ? { ...o, [key]: options[key].default } : o,
+        {}
+      ),
+      help: false
+    }
+  });
+
+  parseDebug(JSON.stringify(argMap));
+
+  const result: any = { ...argMap, help: argMap.help, args: argMap._ };
```

Comment:
`: any` 👀

Reply:
A certain amount of dynamism is required here because we're doing some hacking around adding types to things that actually have type `any` (stuff that comes back from `minimist`). `any` is the most succinct way to represent that, but I've tried to patch out most of the any's with some smarter types in the latest commits.